kratos/utilities/parallel_distance_calculator.cpp

```diff
@@ -0,0 +1,630 @@
+
+// Project includes
+#include "includes/model_part.h"
+#include "utilities/geometry_utilities.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/parallel_distance_calculator.h"
+
+namespace Kratos
+{
+
+template <unsigned int TDim>
+const Kratos::Flags ParallelDistanceCalculator<TDim>::CALCULATE_EXACT_DISTANCES_TO_PLANE(Kratos::Flags::Create(0));
+
+template<unsigned int TDim>
+void ParallelDistanceCalculator<TDim>::CalculateDistances(
+    ModelPart& rModelPart,
+    const Variable<double>& rDistanceVar,
+    const Variable<double>& rAreaVar,
+    const unsigned int MaxLevels,
+    const double MaxDistance,
+    Flags Options)
+{
+    KRATOS_TRY
+
+    Check(rModelPart, rDistanceVar, rAreaVar);
+
+    ResetVariables(rModelPart,rDistanceVar, MaxDistance);
+
+    CalculateExactDistancesOnDividedElements(rModelPart, rDistanceVar, rAreaVar, MaxDistance, Options);
+
+    ExtendDistancesByLayer(rModelPart, rDistanceVar, rAreaVar, MaxLevels, MaxDistance);
+
+    AssignDistanceSign(rModelPart, rDistanceVar, rAreaVar, MaxDistance);
+
+    KRATOS_CATCH("")
+}
+
+template<unsigned int TDim>
+void ParallelDistanceCalculator<TDim>::CalculateInterfacePreservingDistances(
+    ModelPart& rModelPart,
+    const Variable<double>& rDistanceVar,
+    const Variable<double>& rAreaVar,
+    const unsigned int MaxLevels,
+    const double MaxDistance)
+{
+    KRATOS_TRY
+
+    Check(rModelPart, rDistanceVar, rAreaVar);
+
+    ResetVariables(rModelPart,rDistanceVar, MaxDistance);
+
+    AbsDistancesOnDividedElements(rModelPart, rDistanceVar, rAreaVar, MaxDistance);
+
+    ExtendDistancesByLayer(rModelPart, rDistanceVar, rAreaVar, MaxLevels, MaxDistance);
+
+    AssignDistanceSign(rModelPart, rDistanceVar, rAreaVar, MaxDistance);
+
+    KRATOS_CATCH("")
+}
+
+template<unsigned int TDim>
+void ParallelDistanceCalculator<TDim>::CalculateDistancesLagrangianSurface(
+    ModelPart& rModelPart,
+    const Variable<double>& rDistanceVar,
+    const Variable<double>& rAreaVar,
+    const unsigned int MaxLevels,
+    const double MaxDistance)
+{
+    KRATOS_TRY
+
+    //check that variables needed are in the model part
+    const bool is_distributed = rModelPart.IsDistributed();
+    KRATOS_ERROR_IF_NOT(rModelPart.HasNodalSolutionStepVariable(rDistanceVar)) << "Distance variable is not in the model part" << std::endl;
+    KRATOS_ERROR_IF_NOT(rModelPart.HasNodalSolutionStepVariable(rAreaVar)) << "Area Variable is not in the model part" << std::endl;
+    if(is_distributed)
+        KRATOS_ERROR_IF_NOT(rModelPart.HasNodalSolutionStepVariable(PARTITION_INDEX)) << "PARTITION_INDEX variable is not in the model part" << std::endl;
+
+    // set to zero the distance
+    block_for_each(rModelPart.Nodes(), [&](NodeType& rNode){
+        double& r_area = rNode.FastGetSolutionStepValue(rAreaVar);
+        double& r_distance = rNode.FastGetSolutionStepValue(rDistanceVar);
+        rNode.GetValue(rDistanceVar) = r_distance;
+        if (rNode.IsNot(VISITED)) {
```

Comment:
Eventhough you are checking for `VISITED` flag, I could not find where you are setting the flag.